### PR TITLE
Implement JSRPC Connection Hello method

### LINKED ISSUE #3: Communicate with clients via NetString + JSON-RPC
... over the TLS connection as Icinga 2 does.

Handle just the _Hello_ method for now.

### PR DESCRIPTION
**Pull request by yhabteab**
resolves #3

### CODE REVIEW

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Test protocol:

Setup an Icinga cluster where an agent connects to the master.
Turn the master off and this software on instead.
Turn the agent on w/ -x notice.
Are the _Hello_ messages sent+recognized properly?

**Comment by yhabteab** [comment]
Testprotocol:

Start the Server:
```bash
./tls-server -p 5665 -pk /var/lib/icinga2/certs/ha-master.key -puk  /var/lib/icinga2/certs/ha-master.crt -ck  /var/lib/icinga2/ca/ca.crt
```

Start Agent w/ -x notice:
```bash
[2020-12-09 11:32:58 +0100] information/ApiListener: Finished sending replay log for endpoint 'ha-master' in zone 'ha-master'.
[2020-12-09 11:32:58 +0100] information/ApiListener: Finished syncing endpoint 'ha-master' in zone 'ha-master'.
[2020-12-09 11:32:58 +0100] information/ApiListener: Finished reconnecting to endpoint 'ha-master' via host 'localhost' and port '443'
[2020-12-09 11:32:58 +0100] notice/JsonRpcConnection: Received 'icinga::Hello' message from identity 'ha-master'.
[2020-12-09 11:32:58 +0100] notice/JsonRpcConnection: Received 'pki::RequestCertificate' message from identity 'ha-master'.
[2020-12-09 11:32:58 +0100] information/JsonRpcConnection: Received certificate request for CN 'ha-master' signed by our CA.
[2020-12-09 11:32:58 +0100] information/JsonRpcConnection: The certificate for CN 'ha-master' is valid and uptodate. Skipping automated renewal.
[2020-12-09 11:33:03 +0100] notice/CheckerComponent: Pending checkables: 0; Idle checkables: 0; Checks/s: 0
[2020-12-09 11:33:07 +0100] notice/JsonRpcConnection: Error while reading JSON-RPC message for identity 'ha-master': Error: stream truncated
```

Server log:
```bash
2020/12/09 11:32:54 SERVER: Started listening on '0.0.0.0:443'
2020/12/09 11:32:58 SERVER: Accepted connection from [::1]:50585
2020/12/09 11:32:58 SERVER: Connection approved from Server
2020/12/09 11:32:58 SERVER: Connection: Waiting for the client
2020/12/09 11:32:58 SERVER: JSONRPCConnection: Received "icinga::Hello" from the client
2020/12/09 11:32:58 SERVER: Connection: Sending 24 bytes message back
2020/12/09 11:32:58 SERVER: Connection: Waiting for the client
```

**Comment by Al2Klimov** [comment]
Where's the "Received JSRPC Hello from the client" in the logs?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
You've amended your changes on a cherry-picked commit. Undo that.

File: main.go [hunk 1]
```diff
@@ -84,19 +96,31 @@ func main() {
...
+
+		if decoder.Method == "icinga::Hello" {
+			log.Printf("SERVER: JsonRpcConnection: Received %q from the client\n", decoder.Method)
+		}
+
+		err = base.WriteNetStringToStream(conn, n)
 
-		n, err = conn.Write(buf[:n])
-		log.Printf("SERVER: Connection: Sending %d bytes", n)
+		log.Printf("SERVER: Connection: Sending %d bytes message back", unsafe.Sizeof(n))
```
Comment (Al2Klimov) [inline main.go]: Is this how Icinga would handle that method?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go (hunk 1 above)
Comment (yhabteab) [reply main.go]: No, but now it's all about receiving.

Comment (Al2Klimov) [reply main.go]: What's all about receiving? Icinga's _Hello_ handling? Yes. Yours? No.

Comment (Al2Klimov) [inline main.go]: Why `unsafe.Sizeof()`?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go (hunk 1 above)
Comment (yhabteab) [reply main.go]: To find out the size of the outgoing message.

Comment (Al2Klimov) [reply main.go]: Aren't there alternatives for this?

Comment (yhabteab) [reply main.go]: `reflect.TypeOf(variable).Size()` ?

Comment (Al2Klimov) [reply main.go]: Where did you get all these examples from?

Comment (Al2Klimov) [reply main.go]: Your n is a `[]byte`. A quick search for `go size of []byte` shows `len()`.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Squash **your** commits after reordering them on top of the cherry-picked ones via `git rebase -i master`.

**Pull request closed without merging**